fedimint-api/src/core.rs

```diff
@@ -4,38 +4,95 @@
 //!
 //! This (Rust) module defines common interoperability types
 //! and functionality that is used on both client and sever side.
+use core::fmt;
 use std::any::Any;
+use std::borrow::Cow;
 use std::fmt::{Debug, Display};
-use std::io;
 use std::io::Read;
+use std::sync::Arc;
+use std::{io, ops};
 
 pub use bitcoin::KeyPair;
 use fedimint_api::{
-    dyn_newtype_define, dyn_newtype_impl_dyn_clone_passhthrough,
+    dyn_newtype_define,
     encoding::{Decodable, DecodeError, DynEncodable, Encodable},
 };
+use serde::{Deserialize, Serialize};
 
-use crate::ModuleDecoderRegistry;
+use crate::{
+    dyn_newtype_define_with_instance_id, dyn_newtype_impl_dyn_clone_passhthrough_with_instance_id,
+    ModuleDecoderRegistry,
+};
 
 pub mod encode;
 
 pub mod client;
 pub mod server;
 
-/// A module key identifing a module
+/// Module instance ID
 ///
-/// Used as an unique ID, and also as prefix in serialization
-/// of module-specific data.
-pub type ModuleKey = u16;
-
-/// Temporary constant for the modules we already have
+/// This value uniquely identifies a single instance of a module in a federation.
+///
+/// In case a single [`ModuleKind`] is instantiated twice (rare, but possible),
+/// each instance will have a different id.
 ///
-/// To be removed after modularization is complete.
-pub const MODULE_KEY_WALLET: u16 = 0;
-pub const MODULE_KEY_MINT: u16 = 1;
-pub const MODULE_KEY_LN: u16 = 2;
-// not really a module
-pub const MODULE_KEY_GLOBAL: u16 = 1024;
+/// Note: We have used this type differently before, assuming each `u16`
+/// uniquly identifies a type of module in question. This function will move
+/// to a `ModuleKind` type which only identifies type of a module (mint vs wallet vs ln, etc)
+// TODO: turn in a newtype
+pub type ModuleInstanceId = u16;
+
+/// Special ID we use for global dkg
+pub const MODULE_INSTANCE_ID_GLOBAL: u16 = u16::MAX;
+
+// Note: needs to be in alphabetical order of ModuleKind of each module,
+// as this is the ordering we currently harcoded.
+// Should be used only for pre-modularization code we still have  left
+pub const LEGACY_HARDCODED_INSTANCE_ID_LN: ModuleInstanceId = 0;
+pub const LEGACY_HARDCODED_INSTANCE_ID_MINT: ModuleInstanceId = 1;
+pub const LEGACY_HARDCODED_INSTANCE_ID_WALLET: ModuleInstanceId = 2;
+
+/// A type of a module
+///
+/// This is a short string that identifies type of a module.
+/// Authors of 3rd party modules are free to come up with a string,
+/// long enough to avoid conflicts with similiar modules.
+#[derive(Debug, PartialEq, Eq, Clone, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct ModuleKind(Cow<'static, str>);
```

Comment:
~I would have probably avoided that `Cow`, it doesn't seem to be on a critical path and clones are cheap. But no need to change now, complexity impact is still low.~ Associated const requires it.

Reply:
99% of time it will be &'static const, cheap to clone and pass around. But once in a while we read it from a config or something like that.